devito/data/data.py

```diff
@@ -132,16 +140,76 @@ def _global(self, glb_idx, decomposition):
         ret._is_distributed = any(i is not None for i in decomposition)
         return ret
 
+    def check_slicing(func):
+        """Check if the operation will communication across MPI ranks."""
+        @wraps(func)
+        def wrapper(data, *args, **kwargs):
+            glb_idx = args[0]
+            if len(args) > 1 and isinstance(args[1], Data) \
+                    and args[1]._is_mpi_distributed:
+                mpi_slicing = True
+            elif data._is_mpi_distributed:
+                for i in as_tuple(glb_idx):
+                    if isinstance(i, slice) and i.step is not None and i.step < 0:
+                        mpi_slicing = True
+                        break
+                    else:
+                        mpi_slicing = False
+            else:
+                mpi_slicing = False
+            kwargs['mpi_slicing'] = mpi_slicing
+            return func(data, *args, **kwargs)
+        return wrapper
+
     @property
     def _is_mpi_distributed(self):
         return self._is_distributed and configuration['mpi']
 
     def __repr__(self):
         return super(Data, self._local).__repr__()
 
-    def __getitem__(self, glb_idx):
+    @check_slicing
+    def __getitem__(self, glb_idx, mpi_slicing=False):
```

Comment:
instead of ``mpi_slicing`` I'd use ``mode`` (being a string such as ''mpi-slicing", or better an object of type ``Tag``). This would be easier to generalise, I suppose